Tools/clinic/clinic.py

```diff
@@ -5186,111 +5258,55 @@ def state_modulename_name(self, line: str) -> None:
-                name, legacy, kwargs = self.parse_converter(function_node.returns)
-                if legacy:
-                    fail(f"Legacy converter {name!r} not allowed as a return converter")
-                if name not in return_converters:
-                    fail(f"No available return converter called {name!r}")
-                return_converter = return_converters[name](**kwargs)
-            except ValueError:
-                fail(f"Badly formed annotation for {full_name!r}: {returns!r}")
+        return_converter = self.resolve_return_converter(full_name, returns)
 
         fields = [x.strip() for x in full_name.split('.')]
         function_name = fields.pop()
         module, cls = self.clinic._module_and_class(fields)
 
-        if self.kind in {GETTER, SETTER}:
-            if not cls:
-                fail("@getter and @setter must be methods")
+        func = Function(
+            name=function_name,
+            full_name=full_name,
+            module=module,
+            cls=cls,
+            c_basename=c_basename,
+            return_converter=return_converter,
+            kind=self.kind,
+            coexist=self.coexist,
+            critical_section=self.critical_section,
+            target_critical_section=self.target_critical_section
+        )
+        self.add_function(func)
 
-        self.update_function_kind(full_name)
-        if self.kind is METHOD_INIT and not return_converter:
-            return_converter = init_return_converter()
-
-        if not return_converter:
-            return_converter = CReturnConverter()
-
-        self.function = Function(name=function_name, full_name=full_name, module=module, cls=cls, c_basename=c_basename,
-                                 return_converter=return_converter, kind=self.kind, coexist=self.coexist,
-                                 critical_section=self.critical_section,
-                                 target_critical_section=self.target_critical_section)
-        self.block.signatures.append(self.function)
-
-        # insert a self converter automatically
-        type, name = correct_name_for_self(self.function)
-        kwargs = {}
-        if cls and type == "PyObject *":
-            kwargs['type'] = cls.typedef
-        sc = self.function.self_converter = self_converter(name, name, self.function, **kwargs)
-        p_self = Parameter(name, inspect.Parameter.POSITIONAL_ONLY,
-                           function=self.function, converter=sc)
-        self.function.parameters[name] = p_self
-
-        (cls or module).functions.append(self.function)
         self.next(self.state_parameters_start)
 
+    def add_function(self, func: Function) -> None:
+        # Insert a self converter automatically.
+        tp, name = correct_name_for_self(func)
+        kwargs: dict[str, Any] = {}
+        if func.cls and tp == "PyObject *":
+            kwargs['type'] = func.cls.typedef
+        func.self_converter = self_converter(name, name, func, **kwargs)
```

Comment:
This works fine, but you could also consider doing this:

```suggestion
        if func.cls and tp == "PyObject *":
            func.self_converter = self_converter(name, name, func, type=func.cls.typedef)
        else:
            func.self_converter = self_converter(name, name, func)
```

It doesn't end up being any more lines and, because you avoid having to type the values of the kwargs dict as `Any`, it means mypy can check that you are actually supplying the right type for the specific keyword you're supplying here :)

Reply:
Yep, that's better; thanks! Are fine with the rest of the PR?

Reply:
> Are fine with the rest of the PR?

Haven't looked at the rest, I just looked at this because you pinged me 😄 I can look if you'd like me to, though probably not _right_ now (need to go to 🛏️)

Reply:
```suggestion
        if func.cls and tp == "PyObject *":
            func.self_converter = self_converter(name, name, func,
                                                 type=func.cls.typedef)
        else:
            func.self_converter = self_converter(name, name, func)
```